src/Elliptic/BoundaryConditions/AnalyticSolution.hpp

```diff
@@ -0,0 +1,234 @@
+          typename Registrars>
+struct AnalyticSolution;
+
+namespace Registrars {
+template <typename System, size_t Dim = System::volume_dim,
+          typename FieldTags = typename System::primal_fields,
+          typename FluxTags = typename System::primal_fluxes>
+struct AnalyticSolution {
+  template <typename Registrars>
+  using f = BoundaryConditions::AnalyticSolution<System, Dim, FieldTags,
+                                                 FluxTags, Registrars>;
+};
+}  // namespace Registrars
+
+template <typename System, size_t Dim = System::volume_dim,
+          typename FieldTags = typename System::primal_fields,
+          typename FluxTags = typename System::primal_fluxes,
+          typename Registrars =
+              tmpl::list<BoundaryConditions::Registrars::AnalyticSolution<
+                  System, Dim, FieldTags, FluxTags>>>
+struct AnalyticSolution;
+/// \endcond
+
+/*!
+ * \brief Impose the analytic solution on the boundary. Works only if an
+ * analytic solution exists.
+ *
+ * The analytic solution is retrieved from `::Tags::AnalyticSolutionsBase`. It
+ * must hold solutions for both the `System::primal_fields` and the
+ * `System::primal_fluxes`. The user can select to impose the analytic solution
+ * as Dirichlet or Neumann boundary conditions for each field separately.
+ * Dirichlet boundary conditions are imposed on the fields and Neumann boundary
+ * conditions are imposed on the fluxes.
+ *
+ * See `elliptic::Actions::InitializeAnalyticSolutions` for an action that can
+ * add the analytic solutions to the DataBox.
+ */
+template <typename System, size_t Dim, typename... FieldTags,
+          typename... FluxTags, typename Registrars>
+class AnalyticSolution<System, Dim, tmpl::list<FieldTags...>,
+                       tmpl::list<FluxTags...>, Registrars>
+    : public BoundaryCondition<Dim, Registrars> {
+ private:
+  using Base = BoundaryCondition<Dim, Registrars>;
+
+ public:
+  using options =
+      tmpl::list<elliptic::OptionTags::BoundaryConditionType<FieldTags>...>;
+  static constexpr Options::String help =
+      "Boundary conditions from the analytic solution";
+
+  AnalyticSolution() = default;
+  AnalyticSolution(const AnalyticSolution&) noexcept = default;
+  AnalyticSolution& operator=(const AnalyticSolution&) noexcept = default;
+  AnalyticSolution(AnalyticSolution&&) noexcept = default;
+  AnalyticSolution& operator=(AnalyticSolution&&) noexcept = default;
+  ~AnalyticSolution() noexcept = default;
+
+  /// \cond
+  explicit AnalyticSolution(CkMigrateMessage* m) noexcept : Base(m) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(AnalyticSolution);
+  /// \endcond
+
+  /// Select which `elliptic::BoundaryConditionType` to apply for each field
+  explicit AnalyticSolution(
+      const typename elliptic::OptionTags::BoundaryConditionType<
+          FieldTags>::type... boundary_condition_types) noexcept
```

Comment:
I would find this a lot easier to read if you wrote out the type. Does that not work because pack expansion? If that's the case, could you at least document what the type is?

Reply:
Yea I need to have something to expand. Added a comment.